sql/core/src/main/scala/org/apache/spark/sql/ArrowConverters.scala

```diff
@@ -0,0 +1,360 @@
+      fieldNodes.toList.asJava, buffers.toList.asJava)
+
+    buffers.foreach(_.release())
+    recordBatch
+  }
+
+  /**
+   * Write a Field from array of InternalRow to an ArrowBuf.
+   */
+  private def internalRowToArrowBuf(
+      rows: Array[InternalRow],
+      ordinal: Int,
+      field: StructField,
+      allocator: RootAllocator): (Array[ArrowFieldNode], Array[ArrowBuf]) = {
+    val numOfRows = rows.length
+    val columnWriter = ColumnWriter(allocator, field.dataType)
+    columnWriter.init(numOfRows)
+    var index = 0
+
+    while(index < numOfRows) {
+      val row = rows(index)
+      if (row.isNullAt(ordinal)) {
+        columnWriter.writeNull()
+      } else {
+        columnWriter.write(row, ordinal)
+      }
+      index += 1
+    }
+
+    val (arrowFieldNodes, arrowBufs) = columnWriter.finish()
+    (arrowFieldNodes.toArray, arrowBufs.toArray)
+  }
+
+  /**
+   * Convert a Spark Dataset schema to Arrow schema.
+   */
+  private[sql] def schemaToArrowSchema(schema: StructType): Schema = {
+    val arrowFields = schema.fields.map { f =>
+      new Field(f.name, f.nullable, sparkTypeToArrowType(f.dataType), List.empty[Field].asJava)
+    }
+    new Schema(arrowFields.toList.asJava)
+  }
+
+  /**
+   * Write an ArrowPayload to a byte array
+   */
+  private[sql] def payloadToByteArray(payload: ArrowPayload, schema: StructType): Array[Byte] = {
+    val arrowSchema = ArrowConverters.schemaToArrowSchema(schema)
+    val out = new ByteArrayOutputStream()
+    val writer = new ArrowWriter(Channels.newChannel(out), arrowSchema)
+    try {
+      payload.foreach(writer.writeRecordBatch)
+    } catch {
+      case e: Exception =>
+        throw e
+    } finally {
+      writer.close()
+      payload.foreach(_.close())
+    }
+    out.toByteArray
+  }
+}
+
+private[sql] trait ColumnWriter {
+  def init(initialSize: Int): Unit
+  def writeNull(): Unit
+  def write(row: InternalRow, ordinal: Int): Unit
+
+  /**
+   * Clear the column writer and return the ArrowFieldNode and ArrowBuf.
+   * This should be called only once after all the data is written.
+   */
+  def finish(): (Seq[ArrowFieldNode], Seq[ArrowBuf])
+}
+
+/**
+ * Base class for flat arrow column writer, i.e., column without children.
+ */
+private[sql] abstract class PrimitiveColumnWriter(protected val allocator: BaseAllocator)
+    extends ColumnWriter {
+  protected def valueVector: BaseDataValueVector
+  protected def valueMutator: BaseMutator
+
+  protected def setNull(): Unit
+  protected def setValue(row: InternalRow, ordinal: Int): Unit
+
+  protected var count = 0
+  protected var nullCount = 0
+
+  override def init(initialSize: Int): Unit = {
+    valueVector.allocateNew()
+  }
+
+  override def writeNull(): Unit = {
+    setNull()
+    nullCount += 1
+    count += 1
+  }
+
+  override def write(row: InternalRow, ordinal: Int): Unit = {
+    setValue(row, ordinal)
+    count += 1
+  }
+
+  override def finish(): (Seq[ArrowFieldNode], Seq[ArrowBuf]) = {
```

Comment:
Do we need to prevent a `ColumnWriter` has been written after `finish` is called?